src/main.rs

```diff
@@ -1311,78 +1312,119 @@ fn parse_manpage_at_path(
-//         except (KeyboardInterrupt, SystemExit):
-//             raise
-//         except:
-//             add_diagnostic('Error parsing %s: %s' % (manpage_path, sys.exc_info()[0]), BRIEF_VERBOSE)
-//             flush_diagnostics(sys.stderr)
-//             traceback.print_exc(file=sys.stderr)
-//         flush_diagnostics(sys.stderr)
-//     print("") #Newline after loop
-//     add_diagnostic("Successfully parsed %d / %d pages" % (successful_count, total_count), BRIEF_VERBOSE)
-//     flush_diagnostics(sys.stderr)
-
 #[derive(Copy, Clone, Debug)]
 struct Progress(pub bool);
 
-// TODO Arg/output types?
+// Output: Result<cmd_name, error>
+// TODO: Result<cmd_name, CompletionsError>
 fn parse_and_output_man_pages(
-    _paths: impl Iterator<Item = PathBuf>,
-    _output_directory: PathBuf,
-    Progress(_show_progress): Progress,
+    paths: &[PathBuf],
+    output_directory: PathBuf,
+    Progress(show_progress): Progress,
+    deroff_only: bool,
+    write_to_stdout: bool,
 ) {
-    unimplemented!();
+    // Get the number of digits in num
+    #[inline(always)]
+    fn num_digits(num: usize) -> usize {
+        (num as f32).log10() as usize + 1
+    }
+
+    let mut paths = paths;
+    paths.sort();
+    paths.dedup();
+
+    let total = paths.len();
+
+    let (mut successful_count, mut index) = (0, 0);
+    // TODO;
+    let mut cmd_name = "".to_owned();
+
+    let padding = num_digits(total);
+
+    let mut last_len: usize = 0;
+
+    if show_progress && !write_to_stdout {
+        println!(
+            "Parsing man pages and writing completions to {:?}",
+            output_directory
+        );
+    }
+
+    // As far as I can tell, we're fully supporting xz / lzma, so the checks here aren't needed
+
+    for manpage_path in paths {
+        index += 1;
+        // foo/bar/gcc.1.gz -> gcc.1.gz
+        // TODO: Expect
+        let man_file_name = manpage_path
+            .file_name()
+            .and_then(|fname| Some(fname.to_string_lossy().to_mut().to_owned()))
+            .expect(&format!(
+                "Failed to get manfile name from {:?}",
+                manpage_path
+            ));
+
+        // gcc.1.gz -> gcc
+        cmd_name = man_file_name
+            .split('.')
+            .next()
+            .and_then(|cmd_name| Some(cmd_name.to_owned()))
+            .expect(&format!(
+                "Failed to get command name from {}",
+                man_file_name
+            ));
+
+        if show_progress {
+            let num_spaces = padding - num_digits(index);
+
+            let progress = format!(
+                "  {0:>1$} / {2} : {3}",
+                index, num_spaces, total, man_file_name,
+            );
+
+            // padded_progress_str = progress_str.ljust(last_progress_string_length)
+            let padded = format!(
+                "{0:<1$}",
+                progress,
+                last_len.checked_sub(progress.len()).unwrap_or(0)
```

Comment:
Do we need to subtract here?

By the way, what you are doing here is known as `saturating_sub`.

Reply:
No, we don't.